lib/reline/line_editor.rb

```diff
@@ -406,110 +313,239 @@ def multiline_off
+      dialog.contents = nil
+      dialog.trap_key = nil
+    end
+  end
+
+  def update_dialogs(key = nil)
+    @dialog_initialzed = true
+    editor_cursor_x, editor_cursor_y = editor_cursor_position
+    @dialogs.each do |dialog|
+      dialog.trap_key = nil
+      update_each_dialog(dialog, editor_cursor_x, editor_cursor_y - screen_scroll_top, key)
+    end
+  end
+
+  def render_finished
+    clear_rendered_lines
+    render_full_content
+  end
+
+  def clear_rendered_lines
+    Reline::IOGate.move_cursor_up @cursor_y
+    Reline::IOGate.move_cursor_column 0
+
+    num_lines = @rendered_screen_cache&.size
+    return unless num_lines && num_lines >= 1
+
+    Reline::IOGate.move_cursor_down num_lines - 1
+    (num_lines - 1).times do
+      Reline::IOGate.erase_after_cursor
+      Reline::IOGate.move_cursor_up 1
+    end
+    Reline::IOGate.erase_after_cursor
+    @rendered_screen_cache = nil
+  end
+
+  def render_full_content
+    lines = @buffer_of_lines.size.times.map do |i|
+      line = prompt_list[i] + modified_lines[i]
+      wrapped_lines, = split_by_width(line, screen_width)
+      wrapped_lines.last.empty? ? "#{line} " : line
+    end
+    @output.puts lines.map { |l| "#{l}\r\n" }.join
+  end
+
+  def print_nomultiline_prompt(prompt)
+    return unless prompt && !@is_multiline
+
+    # Readline's test `TestRelineAsReadline#test_readline` requires first output to be prompt, not cursor reset escape sequence.
+    @rendered_screen_cache = [[[0, Reline::Unicode.calculate_width(prompt, true), prompt]]]
+    @output.write prompt
+  end
+
+  def render_differential
+    unless @dialog_initialzed
+      update_dialogs
+    end
+
+    editor_cursor_x, editor_cursor_y = editor_cursor_position
+
+    rendered_lines = @rendered_screen_cache || []
+    new_lines = wrapped_lines.flatten[screen_scroll_top, screen_height].map do |l|
+      [[0, Reline::Unicode.calculate_width(l, true), l]]
+    end
+    if @menu_info
+      @menu_info.list.sort!.each do |item|
+        new_lines << [[0, Reline::Unicode.calculate_width(item), item]]
       end
-      @buffer_of_lines[@line_index] = @line
-      @rest_height = 0 if @scroll_partial_screen
-    else
-      line = modify_lines(whole_lines)[@line_index]
-      render_partial(prompt, prompt_width, line, 0)
-      if finished?
-        scroll_down(1)
-        Reline::IOGate.move_cursor_column(0)
-        Reline::IOGate.erase_after_cursor
+      @menu_info = nil # TODO: do not change state here
+    end
+
+    @dialogs.each_with_index do |dialog, index|
+      next unless dialog.contents
+
+      x_range, y_range = dialog_range dialog, editor_cursor_y - screen_scroll_top
+      y_range.each do |row|
+        next if row < 0 || row >= screen_height
+        dialog_rows = new_lines[row] ||= []
+        dialog_rows[index + 1] = [x_range.begin, dialog.width, dialog.contents[row - y_range.begin]]
+      end
+    end
+
+    if new_lines != rendered_lines
+      Reline::IOGate.hide_cursor
```

Comment:
When lines to be rendered changed, reline needs to move cursor and write texts many times. We need to hide cursor because we don't want to show the cursor moving while rendering.
Example of not hiding cursor (Repeating Insert `)` and delete)
![no_hide](https://github.com/ruby/reline/assets/1780201/0df69ffa-8a6e-40a7-9017-0c2c50ba8ef7)

On the other hand, if we always hide cursor in render_differential, cursor sometimes blink when just moving cursor.
Example of pressing arrow keys with hiding cursor
![always_hide](https://github.com/ruby/reline/assets/1780201/71b81f48-6fcd-4eec-b976-bc9d501f22c4)


We need this for comfortable editing, so we don't have to hide cursor in exceptional case like clear screen with `C-l`.